1-Draft/RFC00XX-Module-Isolation.md

```diff
@@ -0,0 +1,251 @@
+  and restored to the previous context after the invocation.
+
+There is risk if we completely depend on the `AssemblyLoadContext.CurrentContextualReflectionContext` to get the active ALC to work with,
+because that property can be changed by a user.
+If `CurrentContextualReflectionContext` is changed to an arbitrary ALC,
+we may run into weird type/assembly resolution issues.
+
+So, I propose to add the new internal property `ExecutionContext.ActiveReflectionContext`,
+which points to the currently active ALC that is supposed to be visible and usable in PowerShell.
+If its value is `null`, then the currently active ALC is the default ALC.
+
+Similarly, the setup and restore of this property should be done before and after command invocation:
+- `ExecutionContext.ActiveReflectionContext` is set before invoking a module command,
+  and restored to the previous context afterwards.
+- `ExecutionContext.ActiveReflectionContext` is set before running a command or script block in a module context (`& $module <cmd-or-script>`),
+  and restored to the previous context after the invocation.
+
+### Update Type Resolution Logic
+
+Now that we have custom ALC's for modules, the type resolving logic needs to be changed:
+
+1. We should first look for the type within the `ExecutionContext.ActiveReflectionContext`,
+and then revert back to the default ALC.
+1. When reverting back to the default ALC,
+we should ignore the assemblies that are loaded (and thus already searched) from the active ALC.
+
+Since different versions of the same assemblies can be loaded into a custom ALC in PowerShell,
+we may have type identity issues within a session -- type `C` and `C'` are the same type in terms of the fully-qualified type name,
+but they are from two different assembly instances that are loaded into two load contexts.
+
+In order to avoid type identity issue in type resolution and have it work consistently,
+we need to do the following extra updates:
+
+1. Today, we have a `TypeCache` data structure that maps a `Tuple<ITypeName, TypeResolutionState>` to a `Type`.
+We will need to have separate `TypeCache` defined in the `PSModuleLoadContext` class,
+so every custom ALC instance has its own type cache.
+The existing `TypeCache` will become the type cache for the default ALC.
+
+1. Today, we have `ExecutionContext.AssemblyCache` that holds all assemblies that are explicitly loaded by PowerShell.
+Simiarly, we will need to have separate `AssemblyCache` defined in the `PSModuleLoadContext` class,
+so every custom ALC instance has its own `AssemblyCache`.
+The existing `AssemblyCache` will be for the default ALC exclusively.
+
+1. All the `ITypeName` implementation types supports type caching.
+The cached type of a `ITypeName` instance should be able to invalid the caching when `ExecutionContext.ActiveReflectionContext` changes.
+
+## Potential Problems
+
+### Module Interoperability
+
```

Comment:
Are there potential problems with serialization in this environment? Currently we don't capture version information about the assemblies which are providing the types that are serialized. If a user takes an object out of an isolated ALC and then serializes it via `export-CLIXL` how will they be able to deserialize it? (especially if the type has changed dramatically)

Reply:
I don't think there will be any potential problems with serialization/deserialization. The deserialization will produce a `PSObject`, not really coupling with a specific version of a type.

For example, say we have a `FileInfo` object out of a custom ALC, and serialize it with `export-clixml`. The object produced by `import-clixml` from the serialization xml file is a `PSObject`, with the type name `Deserialized.System.IO.FileInfo`.

Reply:
We do re-hydrate some core types by constructing that type (Microsoft.PowerShell.DeserializingTypeConverter), and we provide support for custom type re-hydration.  This construction may possibly create the 'wrong' type, but may just be something that a custom type converter needs to be aware of.  But I don't think this is something to prevent us from pursuing ALCs.